xenonpy/contrib/ismd/reactant_pool.py

```diff
@@ -0,0 +1,157 @@
+#  Copyright (c) 2019. yoshida-lab. All rights reserved.
+#  Use of this source code is governed by a BSD-style
+#  license that can be found in the LICENSE file.
+# -*- coding: utf-8 -*-
+
+from xenonpy.inverse.base import BaseProposal, ProposalError
+import random
+import pandas as pd
+
+
+class ReactantNotInPoolError(ProposalError):
+
+    def __init__(self, r_id):
+        super().__init__("reactant id {} is not in the reactant pool".format(r_id))
+
+
+class NotSquareError(ProposalError):
+
+    def __init__(self, n_row=0, n_col=0):
+        super().__init__("dataframe of shape {} * {} is not square".format(n_row, n_col))
+
+
+class SimPoolnotmatchError(ProposalError):
+
+    def __init__(self, n_pool=0, n_sim=0):
+        super().__init__(
+            "reactant pool with length {} dose not match the size of similarity matrix of size {} * {}".format(
+                n_pool, n_sim, n_sim))
+
+
+class NoSampleError(ProposalError):
+
+    def __init__(self):
+        super().__init__("sample is empty")
+
+
+class ReactantPool(BaseProposal):
+
+    def __init__(
+            self,
+            *,
+            pool_df=None,
+            sim_df=None,
+            reactor=None,
+            pool_smiles_col='SMILES',
+            sim_id_in_pool=None,
+            sample_reactant_idx_col='reactant_idx',
+            #sample_reactant_idx_old_col='reactant_idx_old',
+            sample_reactant_smiles_col='reactant_smiles',
+            sample_product_smiles_col='product_smiles',
+            splitter='.'):
+        """
+        A module consists the reactant pool for proposal
+        ----------
+        Parameters:
+            pool_df : a dataframe contains the all usable reactants,
+            sim_df : a matrix of similarity between each pair of reactant in pool_df, size len(pool_df)*len(pool_df).
+            reactor : reaction prediction model.
+            pool_smiles_col : the column name of reactant SMILES in the pool_df,
+            sim_id_in_pool : the column name of id in pool_df corresponding to sim_df, if None, use index.
+            sample_reactant_idx_col : the column name of reactant id in sample dataframe.
+            #sample_reactant_idx_old_col : the column name of old reactant id in sample dataframe, copied from sample_reactant_idx_col before proposal.
+            sample_reactant_smiles_col : the column name of reactant SMILES in sample dataframe
+            sample_product_smiles_col : the column name of product SMILES in sample dataframe
+            splitter : string used for concatenating reactant in a reactant set.
+        """
+        if len(sim_df) != len(sim_df.columns):
+            raise NotSquareError(len(sim_df), len(sim_df.columns))
+
+        if len(pool_df) != len(sim_df):
```

Comment:
If `pool_df` and `sim_df` are very huge, please use `pool_df.shape[0]`  instead of `len(...)` funtion